### PR TITLE
BL-5585 Bloom Features

### PR DESCRIPTION
**Pull request by JohnThomson**
This change is [Reviewable](https://reviewable.io/reviews/bloombooks/bloomreader/114)

### CODE REVIEW

**Comment by JohnThomson** [comment]
This is based on a PR by Rick Conrad. The substantive changes are mostly his, but we decided to take a different approach to what features to look for in the HTML file to control various behaviors more independently.
I have not reversed some code reorganizations that make review difficult, but have tried to comment where it would make things easier.

---

Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions.

---

*[app/src/main/java/org/sil/bloom/reader/BloomFileReader.java, line 107 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7W8njKQ5tqdn4rayIh:-L7W8njKQ5tqdn4rayIi:b-6f7uuh) ([raw file](https://github.com/bloombooks/bloomreader/blob/2beb8d43db82a7bbc54b4e0d3599bae9e80edebf/app/src/main/java/org/sil/bloom/reader/BloomFileReader.java#L107)):*
> ```Java
>     }
> 
>     public String getStringMetaProperty(String property, String defaultIfNotFound){
> ```

This was carried over from Rick's work. As things turned out, I'm not sure whether it is used, but I thought it was worth keeping.

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 174 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7WAbAX3jsrXjSGjK58:-L7WAbAYFYkHHdxh2nDo:b-doc72v) ([raw file](https://github.com/bloombooks/bloomreader/blob/3d194a1110e273316d88cb3eb3c70b33fbd5d38a/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L174)):*
> ```Java
>     // Minimum time a page must be visible before we automatically switch to the next
>     // (if playing audio...usually this only affects pages with no audio)
>     final int MIN_PAGE_SWITCH_MILLIS = 3000;
> ```

moved unchanged

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 183 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7W9x3xYVcSqdoelQgz:-L7W9x3xYVcSqdoelQh-:b7vjz4o) ([raw file](https://github.com/bloombooks/bloomreader/blob/3d194a1110e273316d88cb3eb3c70b33fbd5d38a/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L183)):*
> ```Java
>     // So we do complicated things with a timer to make sure we don't flip the page too soon...
>     // the minimum delay is specified in MIN_PAGE_SWITCH_MILLIS.
>     public void pageAudioCompleted() {
> ```

The way Reviewable is arranging this diff this function appears to have been moved. I think the only change is that instead of always returning, we do so only if mAutoAdvance is false. However, this means that previously unused code is now active, so it may be worth reviewing.

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 210 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7WAJu6Lad1x5PUDu_J:-L7WAJu6Lad1x5PUDu_K:b-doc72v) ([raw file](https://github.com/bloombooks/bloomreader/blob/3d194a1110e273316d88cb3eb3c70b33fbd5d38a/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L210)):*
> ```Java
>     }
> 
>     void clearNextPageTimer() {
> ```

moved unchanged

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 219 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7WATf1ZJu1qGf2ytVm:-L7WATf1ZJu1qGf2ytVn:bd6ange) ([raw file](https://github.com/bloombooks/bloomreader/blob/3d194a1110e273316d88cb3eb3c70b33fbd5d38a/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L219)):*
> ```Java
>     }
> 
>     private void goToNextPageNow() {
> ```

moved unchanged except the multimedia book check was removed.

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 244 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7WAtlktOx28SZe-0cY:-L7WAtlktOx28SZe-0cZ:b-ev20gl) ([raw file](https://github.com/bloombooks/bloomreader/blob/3d194a1110e273316d88cb3eb3c70b33fbd5d38a/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L244)):*
> ```Java
>     }
> 
>     public void narrationPausedChanged() {
> ```

moved unchanged except multimedia check removed.

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 282 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7WB8m9o_YiXiAsFwWo:-L7WB8m9o_YiXiAsFwWp:b14uvdm) ([raw file](https://github.com/bloombooks/bloomreader/blob/3d194a1110e273316d88cb3eb3c70b33fbd5d38a/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L282)):*
> ```Java
> 
>     @Override
>     protected void onNewOrUpdatedBook(String fullPath) {
> ```

these next two also moved unchanged.

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 312 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7W9PtTzNQ5jFfyKRug:-L7W9PtTzNQ5jFfyKRuh:b56cyuk) ([raw file](https://github.com/bloombooks/bloomreader/blob/3d194a1110e273316d88cb3eb3c70b33fbd5d38a/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L312)):*
> ```Java
>     }
> 
>     private void loadBook() {
> ```

This function got inlined, which unfortunately makes the diff hard to follow. Reviewable has happily interpreted it as moving the code above, which is mostly unchanged, and modifying this, which has quite a few changes.

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 316 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7WCQeXu956baYvUBdv:-L7WCQeXu956baYvUBdw:b2hoor1) ([raw file](https://github.com/bloombooks/bloomreader/blob/2beb8d43db82a7bbc54b4e0d3599bae9e80edebf/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L316)):*
> ```Java
>                 mPager.setRotationY(180);
>             mPager.setAdapter(mAdapter);
>             final BloomPageChangeListener listener = new BloomPageChangeListener(audioDirectoryPath);
> ```

What really changed here is that the methods of the previously inline class got moved to an explicit class BloomPageChangeListener. Reviewable has interpreted this as moving the following block of code, which can be seen deleted in several chunks below.

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 400 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7WBseFeL4l2vlICl8g:-L7WBseFeL4l2vlICl8h:bh9wzqk) ([raw file](https://github.com/bloombooks/bloomreader/blob/2beb8d43db82a7bbc54b4e0d3599bae9e80edebf/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L400)):*
> ```Java
>     };
> 
>     // Minimum time a page must be visible before we automatically switch to the next
> ```

A big chunk of moved, not new, code. The various deleted bits parallel to it are really still present above.

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 534 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7WD2HdOw4KNqitgdlN:-L7WD2HdOw4KNqitgdlO:b8189s2) ([raw file](https://github.com/bloombooks/bloomreader/blob/2beb8d43db82a7bbc54b4e0d3599bae9e80edebf/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L534)):*
> ```Java
>     }
> 
>     // Given something like <body data-bfplaymusic="landscape;bloomReader"> for bodyTag
> ```

Now we're past the ugly bit to real changes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-:-L7WDOmvnHKJMcHE4vCh:bdahm8p)*

**Comment by rjconrad** [comment]
:lgtm:  with one question.

---

Review status: 0 of 3 files reviewed at latest revision, 13 unresolved discussions.

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 289 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7Yc9GgE5dOjI-_TXB7:-L7Yc9GgE5dOjI-_TXB8:b-kwpojs) ([raw file](https://github.com/bloombooks/bloomreader/blob/2beb8d43db82a7bbc54b4e0d3599bae9e80edebf/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L289)):*
> ```Java
>                     mBodyTag = "<body>"; // a trivial default, saves messing with nulls.
>                 }
>                 setFeatureEffects();
> ```

Is this superseded by the call to setFeatureEffects() in MakeBrowserForPage()?

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 863 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7Yba20ZnCRnBv6hIgv:-L7Yba20ZnCRnBv6hIgw:bq69ofs) ([raw file](https://github.com/bloombooks/bloomreader/blob/2beb8d43db82a7bbc54b4e0d3599bae9e80edebf/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L863)):*
> ```Java
>                 if(position == 0) {
>                     getPageOrientationAndRotateScreen(page);
>                     setFeatureEffects(); // depends on what orientation we're actually in, so can't do sooner.
> ```

setFeatureEffects is called here when we build the first WebView as well as in the initial book load.  Is the first one superfluous?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-:-L7YcQbzPHBQhx1sOVlE:b-hbp1a3)*

**Comment by JohnThomson** [comment]
Review status: 0 of 3 files reviewed at latest revision, 13 unresolved discussions.

---

*[app/src/main/java/org/sil/bloom/reader/ReaderActivity.java, line 289 at r1](https://reviewable.io:443/reviews/bloombooks/bloomreader/114#-L7Yc9GgE5dOjI-_TXB7:-L7ZlKCEeUrqgrNZqlSt:b-4v0opu) ([raw file](https://github.com/bloombooks/bloomreader/blob/2beb8d43db82a7bbc54b4e0d3599bae9e80edebf/app/src/main/java/org/sil/bloom/reader/ReaderActivity.java#L289)):*
<details><summary><i>Previously, rjconrad (Rick Conrad) wrote…</i></summary><blockquote>

Is this superseded by the call to setFeatureEffects() in MakeBrowserForPage()?
</blockquote></details>

You're right. I removed it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomreader/114)*
